README
Log/Dispatch/File/Alerts version 1.03
=====================================

--------
Abstract
--------

This module subclasses Log::Dispatch::File to implement the generation 
of single-file alerts for log messages. It is intended to be configured 
to recieve only messages of level FATAL. So any fatal error creates a 
single file in the specified directory to call for the help of an 
administrator. When he has finished his work, he can remove the alert 
file to show the problem has been resolved. It allows for a pattern for 
the alert file, so it can generate date/time/pid stamped files. It uses 
flock to ensure being safe in a multitasking environment.

------------
Requirements
------------

This module requires Log::Dispatch::File, included in the Log::Dispatch 
distribution and Log::Log4perl::DateFormat, included in the 
Log::Log4perl distribution. It has only be tested with the versions as 
stated in Makefile.PL, if you find it works with earlier versions, too, 
please send me a email. Thank you.

------------------
Basic Installation
------------------

Log::Dispatch::File::Alerts may be installed through the CPAN shell in 
the usual manner:

  # perl -MCPAN -e 'install Log::Dispatch::File::Alerts'

You can also read this README from the CPAN shell:

  # perl -MCPAN -e shell
  cpan> readme Log::Dispatch::File::Alerts

And you can install the component from the CPAN prompt as well:

  cpan> install Log::Dispatch::File::Alerts

-------------------
Manual Installation
-------------------

Log::Dispatch::File::Alerts can also be installed manually.  The latest 
CPAN version can be found at <http://www.cpan.org/modules/by-
authors/id/J/JA/JACOB/> or in a similarly named directory at your 
favorite CPAN mirror.

Downloading and unpacking the distribution are left as exercises for the 
reader.  To build and test it:

   perl Makefile.PL
   make test

When you're ready to install the component:

  make install

It should now be ready to use.

On Win32 systems, replace "make" in the above commands with "nmake". The 
nmake utility can be downloaded from 
http://download.microsoft.com/download/vc15/Patch/1.52/W95/EN-
US/Nmake15.exe

---------------------
COPYRIGHT AND LICENCE
---------------------

Copyright (C) 2003, 2007, 2010, 2013 M. Jacob <jacob@j-e-b.net>

Based on:
  Log::Dispatch::File::Stamped by Eric Cholet <cholet@logilune.com>
  Log::Dispatch::FileRotate by Mark Pfeiffer, <markpf@mlp-consulting.com.au>

This library is free software; you can redistribute it and/or modify it 
under the same terms as Perl itself. 



Changes
HISTORY
    0.99    Original version; taken from Log::Dispatch::File::Rolling 1.02

    1.00    Initial coding

    1.01    Updated packaging for newer standards. No changes to the coding.

    1.02    Added unlocking of files we do not use.

            Removed the 9999 files limit. Now it will create as many files
            as a Perl integer can support.

    1.03    Adapted to changes in Log::Dispatch::File. If you are useing
            Log::Dispatch::File 2.36 or earlier, use Alerts 1.02.



Makefile.PL
use 5.006001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
	'NAME'          => 'Log::Dispatch::File::Alerts',
	'VERSION_FROM'  => 'lib/Log/Dispatch/File/Alerts.pm', # finds $VERSION
	'PREREQ_PM'     => {
		'Log::Log4perl' => '0.32',
		'Log::Dispatch' => '2.06',
		'Test::More'	=> '1.001002',
#		'Pod::Readme'   => '0.05', # only needed for 'make dist'
	},
	($] >= 5.005 ? (   ## Add these new keywords supported since 5.005
		ABSTRACT_FROM   => 'lib/Log/Dispatch/File/Alerts.pm', # retrieve abstract from module
		AUTHOR          => 'M. Jacob <jacob@j-e-b.net>') : (),
		LICENSE         => 'perl',
		depend          => {
			create_distdir  => 'Changes',
			Changes         => 'lib/Log/Dispatch/File/Alerts.pm'."\n\t".q[$(FULLPERLRUN) -MPod::Readme -e 'Pod::Readme->new( readme_type => "changes" )->parse_from_file( "lib/Log/Dispatch/File/Alerts.pm", "Changes" )']."\n\n",
		},
		realclean       => {
			FILES           => "Changes"
		},
	),
);


lib/Log/Dispatch/File/Alerts.pm
## no critic
package Log::Dispatch::File::Alerts;

use 5.006001;
use strict;
use warnings;

use Log::Dispatch::File '2.37';
use Log::Log4perl::DateFormat;
use Fcntl ':flock'; # import LOCK_* constants

our @ISA = qw(Log::Dispatch::File);

our $VERSION = '1.03';

our $TIME_HIRES_AVAILABLE = undef;

BEGIN { # borrowed from Log::Log4perl::Layout::PatternLayout, Thanks!
	# Check if we've got Time::HiRes. If not, don't make a big fuss,
	# just set a flag so we know later on that we can't have fine-grained
	# time stamps
	
	eval { require Time::HiRes; };
	if ($@) {
		$TIME_HIRES_AVAILABLE = 0;
	} else {
		$TIME_HIRES_AVAILABLE = 1;
	}
}

# Preloaded methods go here.

sub new {
	my $proto = shift;
	my $class = ref $proto || $proto;
	
	my %p = @_;
	
	my $self = bless {}, $class;
	
	# only append mode is supported
	$p{mode} = 'append';
	# 'close' mode is always used
	$p{close_after_write} = 1;
	
	# base class initialization
	$self->_basic_init(%p);

	# split pathname into path, basename, extension
	if ($p{filename} =~ /^(.*)\%d\{([^\}]*)\}(.*)$/) {
		$self->{rolling_filename_prefix}  = $1;
		$self->{rolling_filename_postfix} = $3;
		$self->{rolling_filename_format}  = Log::Log4perl::DateFormat->new($2);
		$self->{filename} = $self->_createFilename(0);
	} elsif ($p{filename} =~ /^(.*)(\.[^\.]+)$/) {
		$self->{rolling_filename_prefix}  = $1;
		$self->{rolling_filename_postfix} = $2;
		$self->{rolling_filename_format}  = Log::Log4perl::DateFormat->new('-yyyy-MM-dd-$!');
		$self->{filename} = $self->_createFilename(0);
	} else {
		$self->{rolling_filename_prefix}  = $p{filename};
		$self->{rolling_filename_postfix} = '';
		$self->{rolling_filename_format}  = Log::Log4perl::DateFormat->new('.yyyy-MM-dd-$!');
		$self->{filename} = $self->_createFilename(0);
	}

	$self->_make_handle();
			
	return $self;
}

sub log_message { # parts borrowed from Log::Dispatch::FileRotate, Thanks!
	my $self = shift;
	my %p = @_;
	my $try = 1;
	my $firstfilename = $self->_createFilename(0); # if this is generated, we are done

	while (defined $try) {
		$self->{filename} = $self->_createFilename($try);

		if (($try > 1 and $firstfilename eq $self->{filename}) or $try < 1) { # later checks for integer overflow
			die 'could not find an unused file for filename "'
			. $self->{filename}
			. '". Did you use "!"?';
		}

		$self->_open_file;
		$self->_lock();
		my $fh = $self->{fh};
		if (not -s $fh) {
			# if the file is zero-sized, it is fresh.
			# else someone else already used it.
			print $fh $p{message};
			$try = undef;
		} else {
			$try++;
		}
		$self->_unlock();
		close($fh);
		$self->{fh} = undef;
	}
}

sub _lock { # borrowed from Log::Dispatch::FileRotate, Thanks!
	my $self = shift;
	flock($self->{fh},LOCK_EX);
	# Make sure we are at the EOF
	seek($self->{fh}, 0, 2);
	return 1;
}

sub _unlock { # borrowed from Log::Dispatch::FileRotate, Thanks!
	my $self = shift;
	flock($self->{fh},LOCK_UN);
	return 1;
}

sub _current_time { # borrowed from Log::Log4perl::Layout::PatternLayout, Thanks!
	# Return secs and optionally msecs if we have Time::HiRes
	if($TIME_HIRES_AVAILABLE) {
		return (Time::HiRes::gettimeofday());
	} else {
		return (time(), 0);
	}
}

sub _createFilename {
	my $self = shift;
	my $try = shift;
	return $self->{rolling_filename_prefix}
	     . $self->_format($try)
	     . $self->{rolling_filename_postfix};
}

sub _format {
	my $self = shift;
	my $try = shift;
	my $result = $self->{rolling_filename_format}->format($self->_current_time());
	$result =~ s/(\$+)/sprintf('%0'.length($1).'.'.length($1).'u', $$)/eg;
	$result =~ s/(\!+)/sprintf('%0'.length($1).'.'.length($1).'u', substr($try, -length($1)))/eg;
	return $result;
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=for changes stop

=head1 NAME

Log::Dispatch::File::Alerts - Object for logging to alert files

=head1 SYNOPSIS

  use Log::Dispatch::File::Alerts;

  my $file = Log::Dispatch::File::Alerts->new(
                             name      => 'file1',
                             min_level => 'emerg',
                             filename  => 'Somefile%d{yyyy!!!!}.log',
                             mode      => 'append' );

  $file->log( level => 'emerg',
              message => "I've fallen and I can't get up\n" );

=head1 ABSTRACT

This module provides an object for logging to files under the
Log::Dispatch::* system.

=head1 DESCRIPTION

This module subclasses Log::Dispatch::File for logging to date/time 
stamped files. See L<Log::Dispatch::File> for instructions on usage. 
This module differs only on the following three points:

=over 4

=item alert files

This module will use a seperate file for every log message.

=item multitasking-safe

This module uses flock() to lock the file while writing to it.

=item stamped filenames

This module supports a special tag in the filename that will expand to 
the current date/time/pid.

It is the same tag Log::Log4perl::Layout::PatternLayout uses, see 
L<Log::Log4perl::Layout::PatternLayout>, chapter "Fine-tune the date". 
In short: Include a "%d{...}" in the filename where "..." is a format 
string according to the SimpleDateFormat in the Java World 
(http://java.sun.com/j2se/1.3/docs/api/java/text/SimpleDateFormat.html). 
See also L<Log::Log4perl::DateFormat> for information about further 
restrictions.

In addition to the format provided by Log::Log4perl::DateFormat this 
module also supports '$' for inserting the PID and '!' for inserting a 
uniq number. Repeat the character to define how many character wide the 
field should be.

A note on the '!': The module first tries to find a fresh filename with this set 
to 1. If there is already a file with that name then it is increased until 
either a free filename has been found. If there is no free filename (e.g. you 
used '!!' and there are already 100 files) or the counter goes over the top 
(integer overflow) the module dies. So if you used many '!'s and there are many 
alert files, this can take quite a while. But if you have that many alert files, 
something already went very bad, so it should not really matter.

=back

=head1 METHODS

=over 4

=item new()

See L<Log::Dispatch::File> and chapter DESCRIPTION above.

=item log_message()

See L<Log::Dispatch::File> and chapter DESCRIPTION above.

=back

=for changes continue

=head1 HISTORY

=over 8

=item 0.99

Original version; taken from Log::Dispatch::File::Rolling 1.02

=item 1.00

Initial coding

=item 1.01

Updated packaging for newer standards. No changes to the coding.

=item 1.02

Added unlocking of files we do not use.

Removed the 9999 files limit. Now it will create as many files as a Perl integer 
can support.

=item 1.03

Adapted to changes in Log::Dispatch::File. If you are useing Log::Dispatch::File
2.36 or earlier, use Alerts 1.02.

=back

=for changes stop

=head1 SEE ALSO

L<Log::Dispatch::File>, L<Log::Log4perl::Layout::PatternLayout>, 
L<Log::Dispatch::File::Rolling>, L<Log::Log4perl::DateFormat>, 
http://java.sun.com/j2se/1.3/docs/api/java/text/SimpleDateFormat.html, 
'perldoc -f flock'

=head1 AUTHOR

M. Jacob, E<lt>jacob@j-e-b.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003, 2007, 2010, 2013 M. Jacob E<lt>jacob@j-e-b.netE<gt>

Based on:

  Log::Dispatch::File::Stamped by Eric Cholet <cholet@logilune.com>
  Log::Dispatch::FileRotate by Mark Pfeiffer, <markpf@mlp-consulting.com.au>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
